File: anb/leetcode/dynamic programming/322. Coin Change.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        int dp[10001]={}, i, j;
        memset(dp, 0x3F, sizeof dp);
        dp[0] = 0;
        for(i=0;i<coins.size();i++) 
            for(j=coins[i];j<=amount;j++)
                dp[j] = min(dp[j], dp[j-coins[i]] + 1);
        return dp[amount] == 0x3F3F3F3F ? -1 : dp[amount];
    }
};

File: 100milliongold/leetcode/dynamic-programming/322.Coin Change.js
/**
 * @param {number[]} coins
 * @param {number} amount
 * @return {number}
 */
 var coinChange = function(coins, amount) {
    const dp = new Array(amount + 1).fill(Infinity);

  // 초기값
  dp[0] = 0;

  for (let money = 1; money <= amount; money++) {
    for (let i = 0; i < coins.length; i++) {
      if (coins[i] <= money) {
        dp[money] = Math.min(dp[money], dp[money - coins[i]] + 1);
      }
    }
  }

  const answer = dp[amount] === Infinity ? -1 : dp[amount];
  return answer;
};

console.log(coinChange([186,419,83,408] ,6249))